packages/amplify_datastore/example/integration_test/relationship_type_test.dart

```diff
@@ -0,0 +1,367 @@
+        await Amplify.DataStore.delete(parent);
+        var parents = await Amplify.DataStore.query(BelongsToModel.classType);
+        expect(parents, isEmpty);
+      });
+
+      testWidgets('delete child', (WidgetTester tester) async {
+        await Amplify.DataStore.delete(child);
+        var children = await Amplify.DataStore.query(ChildModel.classType);
+        expect(children, isEmpty);
+      });
+    });
+
+    group('HasMany', () {
+      var parent = HasManyModel(name: 'HasMany');
+      var children = List.generate(
+          5, (i) => HasManyChildModel(name: 'child $i', parent: parent));
+      late Future<SubscriptionEvent<HasManyChildModel>> childEvent;
+      late Future<SubscriptionEvent<HasManyModel>> hasManyEvent;
+
+      setUpAll(() async {
+        await configureDataStore();
+        await clearDataStore();
+
+        childEvent =
+            Amplify.DataStore.observe(HasManyChildModel.classType).first;
+        hasManyEvent = Amplify.DataStore.observe(HasManyModel.classType).first;
+      });
+
+      testWidgets('precondition', (WidgetTester tester) async {
+        var queriedChildren =
+            await Amplify.DataStore.query(HasManyChildModel.classType);
+        expect(queriedChildren, isEmpty);
+        var queriedParents =
+            await Amplify.DataStore.query(HasManyModel.classType);
+        expect(queriedParents, isEmpty);
+      });
+
+      testWidgets('save parent', (WidgetTester tester) async {
+        await Amplify.DataStore.save(parent);
+        var parents = await Amplify.DataStore.query(HasManyModel.classType);
+        expect(parents, isNotEmpty);
+      });
+
+      testWidgets('save children', (WidgetTester tester) async {
+        for (var child in children) {
+          await Amplify.DataStore.save(child);
+        }
+        var queriedChildren =
+            await Amplify.DataStore.query(HasManyChildModel.classType);
+        expect(queriedChildren, isNotEmpty);
+      });
+
+      testWidgets('query parent', (WidgetTester tester) async {
+        var parents = await Amplify.DataStore.query(HasManyModel.classType);
+        var queriedParent = parents.first;
+        expect(queriedParent, parent);
+        expect(queriedParent.id, parent.id);
+        expect(queriedParent.name, parent.name);
+      });
+
+      testWidgets('query children', (WidgetTester tester) async {
+        var queriedChildren =
+            await Amplify.DataStore.query(HasManyChildModel.classType);
+        for (var i = 0; i < 5; i++) {
+          var queriedChild = queriedChildren[i];
+          var actualChild = children[i];
+          expect(queriedChild, actualChild);
+          expect(queriedChild.parent, actualChild.parent);
+        }
+      });
+
+      testWidgets('observe parent', (WidgetTester tester) async {
+        var event = await hasManyEvent;
+        var observedParent = event.item;
+        expect(observedParent.id, parent.id);
+        expect(observedParent.name, parent.name);
```

Comment:
I updated this to do the equality check on the model. The reason I did not do that originally is because if/when https://github.com/aws-amplify/amplify-flutter/issues/642 is fully resolved, that equality check *may* fail since `parent.children` is null, but `observedParent.children` may eagerly load data (depending on how the issue is resolved).

For now, the full equality check makes sense. I did leave a note that it may need to be updated.